remotes/fixup.go

```diff
@@ -2,88 +2,212 @@ package remotes
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"os"
+	"sync"
 
 	"github.com/containerd/containerd/images"
-	"github.com/containerd/containerd/remotes/docker"
+	"github.com/containerd/containerd/remotes"
 	"github.com/deislabs/duffle/pkg/bundle"
-	"github.com/docker/cli/opts"
+	"github.com/docker/cli/cli/config/configfile"
 	"github.com/docker/distribution/reference"
 	ocischemav1 "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
+const (
+	defaultMaxConcurrentJobs = 4
+	defaultJobsBufferLength  = 50
+)
+
+func noopEventCallback(FixupEvent) {}
+
+// fixupConfig defines the input required for a Fixup operation
+type fixupConfig struct {
+	ctx               context.Context
+	bundle            *bundle.Bundle
+	targetRef         reference.Named
+	eventCallback     func(FixupEvent)
+	maxConcurrentJobs int
+	jobsBufferLength  int
+	resolverConfig    ResolverConfig
+}
+
+func (cfg *fixupConfig) complete() error {
+	if cfg.resolverConfig.resolver == nil || cfg.resolverConfig.originProviderWrapper == nil {
+		return errors.New("resolver and originProviderWrapper are required, please use a complete ResolverConfig")
+	}
+	return nil
+}
+
+// WithContext runs the Fixup operation within a context to handle cancellation
+func WithContext(ctx context.Context) FixupOption {
```

Comment:
The `ctx` should always be the first argument to the function, so `FixupBundle` in this case. If you then want to immediately stash it in `fixupConfig` instead of propagating through the internal call chains then fine (a bit odd, but it's internal so meh). Callers who don't care about contexts should just use `context.Background()` (or `.TODO()`)